kernel/src/engine/arrow_expression.rs

```diff
@@ -67,6 +74,20 @@ impl Scalar {
                     .try_collect()?;
                 Arc::new(StructArray::try_new(fields, arrays, None)?)
             }
+            Array(data) => {
+                let values = data.array_elements();
+                let vecs: Vec<_> = values.iter().map(|v| v.to_array(num_rows)).try_collect()?;
```

Comment:
I'm not 100% confident I understand what this code is doing but:

It looks like we need to replicate an "array scalar" value into an arrow array suitable for comparing against a batch of primitive values, right? So e.g. if I had `x in (1, 2, 3)` then `values` would be `[1, 2, 3]` and we need to explode that into `[[1, 2, 3], [1, 2, 3], ..., [1, 2, 3]]` so that the eventual `evaluate_expression` call can invoke e.g. [in_list](https://arrow.apache.org/rust/arrow/compute/kernels/comparison/fn.in_list.html)? The documentation for that function is sorely incomplete, but I guess it's marching element by element through two arrays, producing a true output element each row whose list array element contains the corresponding primitive array element? That would be general enough to handle a correlated subquery, but would be quite space-inefficient for the common case (literal in-list or uncorrelated subquery) where we compare against the same array in every row -- especially in case said array is large. 

Does arrow rust provide a "scalar" version of `in_list` that takes two primitive arrays instead? So that e.g. `scalar_in_list([1, 2, 3, 4, 5, 6], [1, 3, 5])` returns `[true, false, true, false, true, false]`?

Reply:
You are correct, go through and build the offsets. This approach was actually heavily inspired from how datafusion does this, so while I agree a very large list likely doesn't have great performance there isn't (to my current knowledge) a more idiomatic way to accomplish this. I generally dislike working with the list types in arrow, but to my knowledge arrow does not provide a scalar version of it. You would have to replicate the static value N number of times for each row. Primitive array, Generic string array on the left and list array on the right.

Reply:
I mean really we should be build+probe a hash table for all but the smallest in-lists, to avoid paying `O(n**2)` work. 
If arrow doesn't give a way to do that maybe we need to do it ourselves?

Reply:
Update: the hash table thing gets pretty clearly into engine territory, so if we can figure out how to offload it to engine [data], that's probably better.